File: 117.Populating_Next_Right_Pointers_in_Each_Node_II.cpp

 //Definition for binary tree with next pointer.
struct TreeLinkNode {
   int val;
   TreeLinkNode *left, *right, *next;
   TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 };

class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root) return;
        queue<TreeLinkNode*> q;
        q.push(root);
        int currCnt = 1, nxtCnt = 0;
        while (!q.empty()) {
            TreeLinkNode* peek = q.front();
            q.pop();
            if (--currCnt > 0) {
                peek->next = q.front();
            } else {
                peek->next = NULL;
            }
            if (peek->left) {
                q.push(peek->left);
                nxtCnt++;
            }
            if (peek->right) {
                q.push(peek->right);
                nxtCnt++;
            }
            if (currCnt == 0) {
                currCnt = nxtCnt;
                nxtCnt = 0;
            }
        }
    }
};

File: java/117. Populating Next Right Pointers in Each Node II.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
// O(n) time O(1) space
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null) {
            return;
        }
        TreeLinkNode head = null, prev = null;
        while (root != null || head != null) {
            if (root == null) {
                root = head;
                head = prev = null;
            } else {
                if (root.left != null) {
                    if (head == null) {
                        head = prev = root.left;
                    } else {
                        prev.next = root.left;
                        prev = prev.next;
                    }
                }
                if (root.right != null) {
                    if (head == null) {
                        head = prev = root.right;
                    } else {
                        prev.next = root.right;
                        prev = prev.next;
                    }
                }
                root = root.next;
            }
        }
    }
}
